internal/dns/subenum.go

```diff
@@ -44,23 +64,94 @@ func getSubdomains(ctx context.Context, domain string) ([]string, error) {
 	return subdomains, err
 }
 
-// GetDomainSubdomains queries subfinder for all subdomains for a given domain. It returns a SubdomainsEnumReport struct containing
+// GetDomainSubdomainsBrute queries subfinder for all subdomains for a given domain. It returns a SubdomainsEnumReport struct containing
 // all subdomains and any errors that occurred.
-func GetDomainSubdomains(ctx context.Context, domain string) (SubdomainsEnumReport, error) {
+func GetDomainSubdomainsBrute(ctx context.Context, domain string, subdomainList []string, parallelThreads int, recursiveDepth int, timeout int) (osintscan.DnsSubenumReport, error) {
+	report := osintscan.DnsSubenumReport{
+		Domain:          domain,
+		EnumerationType: osintscan.DnsSubenumTypeBrute,
+	}
 	errors := []string{}
 
-	// 1. Get all valid subdomains
-	subdomains, err := getSubdomains(ctx, domain)
-	if err != nil {
-		errors = append(errors, err.Error())
+	subdomains := getSubdomainsBrute(ctx, domain, subdomainList, parallelThreads, recursiveDepth, timeout)
+
+	report.Subdomains = subdomains
+	report.Errors = errors
+	return report, nil
+
+}
+
+func getSubdomainsBrute(ctx context.Context, domain string, subdomainList []string, parallelThreads int, recursiveDepth int, timeout int) []string {
+	subdomains := []string{}
+	subdomainsSet := make(map[string]struct{}) // To track unique valid subdomains
+	subdomainsMutex := &sync.Mutex{}
+	semaphore := make(chan struct{}, parallelThreads)
+	var wg sync.WaitGroup
+
+	var cancel context.CancelFunc
+	if timeout != 0 {
+		ctx, cancel = context.WithTimeout(ctx, time.Duration(timeout)*time.Minute)
+		defer cancel()
 	}
 
-	// 2. Create report and write to file
-	report := SubdomainsEnumReport{
-		Domain:     domain,
-		Subdomains: subdomains,
-		Errors:     errors,
+	resolver := net.Resolver{}
+
+	// Pre-generate all permutations before starting goroutines
+	allPermutations := make([]string, 0)
```

Comment:
I was thinking when I did this for the interview that doing all permutations gets out of hand from a perf standpoint, so instead could only check at recursive depth n for subdomains that resolve for depth n-1?

E.g., if `mail.google.com` resolves but `fake.google.com` doesn't at depth n=1, then only check the wordlist against `mail.google.com` at n=2.

Reply:
Ya but then you potentially miss subdomainas as stated in the comment. Something worth talking about though because the perms would be extremely high

Reply:
@methodsecuritysj whats you opinion?

Reply:
But if its unusable then the point is moot. I need to see how long it would take to run at a high setting

Reply:
Yeah, extremely high indeed -- all permutations would be factorial at each iteration, which should make us take prohibitively long for any non-trivial wordlist size / depth combination.

I'm not clear on why we miss subdomains if we don't enumerate all the possibilities ahead of time -- like I thought we could generate a new set of goroutines at each recursive depth

Reply:
Yep - that takes way too long. Will update code

Reply:
Makes sense to me, make sure to just put that in the docs